atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java

```diff
@@ -69,157 +74,191 @@
+                logger.accept(log::trace, stopwatch);
+            }
+            if (slowlogger.isWarnEnabled() && shouldSlowLog.test(stopwatch)) {
+                logger.accept(slowlogger::warn, stopwatch);
+            }
+        } else {
+            runnable.run();
+        }
+    }
+
+    private <T> T maybeLog(Supplier<T> supplier, Predicate<Stopwatch> shouldSlowLog, BiConsumer<LoggingFunction, Stopwatch> logger) {
+        if (log.isTraceEnabled() || slowlogger.isWarnEnabled()) {
+            Stopwatch stopwatch = Stopwatch.createStarted();
+            T result = supplier.get();
+            stopwatch.stop();
+            if (log.isTraceEnabled()) {
+                logger.accept(log::trace, stopwatch);
+            }
+            if (slowlogger.isWarnEnabled() && shouldSlowLog.test(stopwatch)) {
+                logger.accept(slowlogger::warn, stopwatch);
+            }
+            return result;
+        } else {
+            return supplier.get();
+        }
+    }
+
+    private <T> T maybeLog(Supplier<T> supplier, Predicate<Stopwatch> shouldSlowLog, TriConsumer<LoggingFunction, Stopwatch, T> logger) {
+        if (log.isTraceEnabled() || slowlogger.isWarnEnabled()) {
+            Stopwatch stopwatch = Stopwatch.createStarted();
+            T result = supplier.get();
+            stopwatch.stop();
+            if (log.isTraceEnabled()) {
+                logger.accept(log::trace, stopwatch, result);
+            }
+            if (slowlogger.isWarnEnabled() && shouldSlowLog.test(stopwatch)) {
+                logger.accept(slowlogger::warn, stopwatch, result);
+            }
+            return result;
+        } else {
+            return supplier.get();
+        }
+    }
+
     private final KeyValueService delegate;
 
-    private ProfilingKeyValueService(KeyValueService delegate) {
+    private final Long slowLogThresholdMillis;
+
+    private final Predicate<Stopwatch> slowLogPredicate;
+
+    private ProfilingKeyValueService(KeyValueService delegate, long slowLogThresholdMillis) {
         this.delegate = delegate;
+        this.slowLogThresholdMillis = slowLogThresholdMillis;
+        slowLogPredicate =  stopwatch -> stopwatch.elapsed(TimeUnit.MILLISECONDS) > slowLogThresholdMillis;
     }
 
     @Override
     public void addGarbageCollectionSentinelValues(TableReference tableRef, Set<Cell> cells) {
-        if (log.isTraceEnabled()) {
-            Stopwatch stopwatch = Stopwatch.createStarted();
-            delegate.addGarbageCollectionSentinelValues(tableRef, cells);
-            log.trace("Call to KVS.addGarbageCollectionSentinelValues on table {} over {} cells took {} ms.",
-                    tableRef, cells.size(), stopwatch.elapsed(TimeUnit.MILLISECONDS));
-        } else {
-            delegate.addGarbageCollectionSentinelValues(tableRef, cells);
-        }
+        maybeLog(() -> delegate.addGarbageCollectionSentinelValues(tableRef, cells),
+                slowLogPredicate,
```

Comment:
seems like this argument is universal to all callers and should instead be used in maybeLog.  Ie. there is no need to inject it each time since we always use the same one.